library/std/src/io/mod.rs

```diff
@@ -452,7 +452,7 @@ pub(crate) fn default_read_to_end<R: Read + ?Sized>(
     let mut consecutive_short_reads = 0;
 
     loop {
-        if buf.len() == buf.capacity() && buf.capacity() == start_cap {
+        if buf.spare_capacity_mut().len() < PROBE_SIZE && buf.capacity() == start_cap {
```

Comment:
I thought that it would be a good idea too, but `small_probe_read` is designed to check for EOF so taking this branch if we still have space in the vector does not seem right.